datadriven/src/sgpp/datadriven/algorithm/GridFactory.hpp

```diff
@@ -31,25 +31,95 @@ class GridFactory {
+
+  /*
+   * calculates hierachical parent interactions
+   * @param vector of resolution
+   * @param defines if only for the first level the hierarchical interactions are created
+   * @param defines if only for the first parent each the hierarchical interactions are created
+   * @return all hierachical interactions of all pixels in a vector
+   */
+  void getHierarchicalParents(std::set<std::set<size_t>>& interactions,
+                              sgpp::datadriven::GeometryConfiguration& geometryConf,
+                              sgpp::datadriven::StencilConfiguration& stencilConf) const;
 
   /*
    * calculates direct neighbour interactions
    * @param vector of resolution
    * @return all direct neighbour interactions of all pixels in a vector
    */
-  std::vector<std::vector<size_t>> getDirectNeighbours(std::vector<int64_t>& res) const;
+  void getDirectNeighbours(std::set<std::set<size_t>>& interactions,
+                           sgpp::datadriven::GeometryConfiguration& geometryConf,
+                           sgpp::datadriven::StencilConfiguration& stencilConf) const;
 
  private:
+  void addChildParentInteraction(
+      std::vector<double>& ratio, size_t numberOfAxes, size_t currentAxis,
+      std::vector<int64_t>& childPosition, std::vector<int64_t>& parentPosition,
+      std::vector<size_t>& parentMultiplicators, std::vector<size_t>& childMultiplicators,
+      size_t parentOffset, size_t childOffset, std::set<std::set<size_t>>& interactions) const;
+
+  /*
+   * calculates the index of a given position. This method exspect a row wise data layout
+   * with each additional dimension beeing stored between the data of the pixels
+   * @param vector containing the size of each dimension
+   * @param vector containing the position
+   * @return the index of the given position in the data
+   */
+  size_t getDataIndex(size_t numberOfDimensions, std::vector<size_t>& multiplicators,
+                      std::vector<int64_t>& position) const;
+
+  /*
+   * Increases a given position like a counter. Therefore if a overflow happens the position will be
+   * set to 0 in all dimensions
+   * @param vector of resolution
+   * @param current position
+   */
+  void getNextPosition(std::vector<int64_t>& dimension, std::vector<int64_t>& position,
+                       size_t colorIndex) const;
+
+  /*
+   * Calculates the multiplicators for each level. The multiplicator is the change in the data index
+   * for one step further into each dimension
+   * @param vector of resolutions
+   * @return vector of vectors containing the multiplicator for each dimension
+   */
+  std::vector<std::vector<size_t>> getMultiplicatorsPerLevel(
+      std::vector<std::vector<int64_t>>& imageDimensions) const;
+
+  void addColorInteractions(std::vector<int64_t>& layerDim, size_t colorIndex, size_t offset,
+                            std::vector<size_t>& multiplicators,
+                            std::set<std::set<size_t>>& interactions) const;
```

Comment:
Without `const`s or Doxygen documentation (`in`/`out` keywords), it's hard to tell if an argument is changed or not.